New Void Endpoint (#436)
centrapay/centrapay-docs: src/api/payment-requests/payment-requests.md

 * **new**: after being created.
 * **paid**: after being paid with one or more transactions.
 * **cancelled**: after being cancelled by the merchant.
 * **expired**: after expiry time is reached without being paid or cancelled.
 * **refunded**: after being partially or fully refunded.

>  * **voided**: after being voided by the merchant.

Author: See explanation [here](https://github.com/centrapay/centrapay-docs/pull/435#discussion_r890638029).

Reviewer: Lets put this back, we've talked this over and it's not a new status on the payment request that we're after. That discussion was around void payment activities being used to support legacy endpoints.

{% endreqspec %}

{% h4 Example response payload %}
{% json %}
{
  "type": "voidCancel",
	"cancellationReason": "CANCELLED_BY_MERCHANT",
  "value": {
    "currency": "NZD",
    "amount": "1000",
  },
  "assetType": "centrapay.nzd.main",
  "paymentRequestId": "MhocUmpxxmgdHjr7DgKoKw",
  "merchantName": "Centrapay Café",
  "merchantId": "26d3Cp3rJmbMHnuNJmks2N",
  "merchantAccountId": "C4QnjXvj8At6SMsEN4LRi9",
  "merchantConfigId": "5efbe2fb96c08357bb2b9242",
  "createdAt": "2021-06-08T04:04:27.426Z",
  "createdBy": "crn::user:0af834c8-1110-11ec-9072-3e22fb52e878",
  "paymentRequestCreatedBy": "crn::user:0af834c8-1110-11ec-9072-3e22fb52e878",
  "activityNumber": "2",
}
{% endjson %}

{% h4 Error Responses %}

| Status |                Code                |                                           Description                                            |
| :----- | :--------------------------------- | :----------------------------------------------------------------------------------------------- |
| 403    | {% break _ VOID_WINDOW_EXCEEDED %} | The time since the payment request was created exceeds the time a payment request can be voided. |

> This endpoint can also return the same error responses as the [Refund][] and Cancel payment request endpoints.

Reviewer: Can we list these error responses out here please?

| 403    | {% break _ REFUND_NOT_SUPPORTED %}          | The asset type does not support refunds.                                                                                                                                                                                                                                    |
| 403    | {% break _ REFUND_WINDOW_EXCEEDED %}        | The time since the payment exceeds the window of time a payment request can be refunded in.                                                                                                                                                                                 |

<a name="void">
### Void a Payment Request **EXPERIMENTAL**

Voiding a payment request will cancel the request and trigger any refunds if necessary.

{% reqspec %}
  POST '/api/payment-requests/{paymentRequestId}/void'
  auth 'jwt'
  path_param 'paymentRequestId', 'MhocUmpxxmgdHjr7DgKoKw'
{% endreqspec %}

{% h4 Example response payload %}
{% json %}
{
  "type": "voidCancel",
  "status": "voided",

> 	"cancellationReason": "CANCELLED_BY_MERCHANT",

Reviewer: Can we get a better example here? This might not fit the bill.

  "value": {
    "currency": "NZD",
    "amount": "1000",
  },
  "assetType": "centrapay.nzd.main",
  "paymentRequestId": "MhocUmpxxmgdHjr7DgKoKw",
  "merchantName": "Centrapay Café",
  "merchantId": "26d3Cp3rJmbMHnuNJmks2N",
  "merchantAccountId": "C4QnjXvj8At6SMsEN4LRi9",
  "merchantConfigId": "5efbe2fb96c08357bb2b9242",
  "createdAt": "2021-06-08T04:04:27.426Z",
  "createdBy": "crn::user:0af834c8-1110-11ec-9072-3e22fb52e878",
  "paymentRequestCreatedBy": "crn::user:0af834c8-1110-11ec-9072-3e22fb52e878",
  "activityNumber": "2",
}
{% endjson %}

{% h4 Error Responses %}

| Status |                Code                |              Description              |
| :----- | :--------------------------------- | :------------------------------------ |

> | 403    | {% break _ VOID_WINDOW_EXCEEDED %} | 24 hours have passed since createdAt. |

Reviewer: > 24 hours have passed since createdAt.

Can we explain this is the void window?

Can we refer to the refund endpoint, someone gets this error they're going to want a fix.

Something like...

> The void window is 24 hours. If you need to reverse this transaction you can use our [Refund][] endpoint.

{% h4 Example response payload %}
{% json %}
{

>   "type": "cancelled",

Author: `cancelled` is not consistent with other payment activity types such as `payment` or `refund`. Therefore I have updated the type to `cancel`.

Reviewer 2: Yea cancelled is for the paymentRequest and cancel is the activity

| 403    | {% break _ REQUEST_EXPIRED %} | The cancellation cannot be completed because the `expiresAt` of the request has passed. |
| 403    | {% break _ REQUEST_PAID %}    | The cancellation cannot be completed because the request `status` is `paid`.            |

<a name="void">
### Void a Payment Request **EXPERIMENTAL**

Voiding a payment request will cancel the request and trigger any refunds if necessary.

{% reqspec %}
  POST '/api/payment-requests/{paymentRequestId}/void'
  auth 'jwt'
  path_param 'paymentRequestId', 'MhocUmpxxmgdHjr7DgKoKw'
{% endreqspec %}

{% h4 Example response payload %}
{% json %}
{
  "type": "voided",
	"cancellationReason": "CANCELLED_BY_MERCHANT",
  "value": {
    "currency": "NZD",
    "amount": "1000",
  },
  "assetType": "centrapay.nzd.main",

>   "paymentRequestId": "MhocUmpxxmgdHjr7DgKoKw",

Reviewer 3: Can you please add the `shortCode` field here? `"shortCode": "CP-C7F-ZS5-032"`

Author: Sure thing 😁

  },
  "assetType": "centrapay.nzd.main",
  "paymentRequestId": "MhocUmpxxmgdHjr7DgKoKw",
  "shortCode": "CP-C7F-ZS5-032",
  "merchantName": "Centrapay Café",
  "merchantId": "26d3Cp3rJmbMHnuNJmks2N",
  "merchantAccountId": "C4QnjXvj8At6SMsEN4LRi9",
  "merchantConfigId": "5efbe2fb96c08357bb2b9242",
  "createdAt": "2021-06-08T04:04:27.426Z",
  "createdBy": "crn::user:0af834c8-1110-11ec-9072-3e22fb52e878",
  "paymentRequestCreatedBy": "crn::user:0af834c8-1110-11ec-9072-3e22fb52e878",
  "activityNumber": "3",
},
{% endjson %}

{% h4 Error Responses %}

| Status |                Code                |                                                                                                                                 Description                                                                                                                                 |
| :----- | :--------------------------------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | 403    | {% break _ VOID_WINDOW_EXCEEDED %} | The void window is closed 24 hours after the request's `createdAt`. If the request is paid, the [Refund][] endpoint can be used to reverse the payment after the void window has closed.                                                                                    |

Reviewer 2: This sentence could be tidied up a bit something like
``The void window lasts for 24 hours after the payment request has been created. If the request has been paid, the [Refund][] endpoint can be used to reverse the payment.``

| 403    | {% break _ ALREADY_REFUNDED %}     | The payment request has already been refunded.                                                                                                                                                                                                                              |

> | 403    | {% break _ REPEAT_REFERENCE %}     | A separate refund request for the payment request has the same external reference. Attempting to refund the payment request twice with the same external reference is not allowed. If the amount of the refund is the same we assume it is a repeat request and return 200. |

Reviewer 2: ```suggestion
| 403    | {% break _ REPEAT_REFERENCE %}     | A refund has already been requested with the same external reference. Refunding the payment request twice with the same external reference is not allowed. If the amount of the refund is the same we assume it is a repeat request and return 200. |
```
